core/src/main/scala/org/apache/spark/util/JsonProtocol.scala

```diff
@@ -69,6 +69,11 @@ private[spark] object JsonProtocol extends JsonUtils {
   private[util]
   val defaultOptions: JsonProtocolOptions = new JsonProtocolOptions(new SparkConf(false))
 
+  // SPARK-49872 remove limit on string lengths
+  mapper.getFactory.setStreamReadConstraints(
+    StreamReadConstraints.builder().maxStringLength(Int.MaxValue).build()
```

Comment:
Two additional limits were introduced in  jackson version 2.15.0. Should we consider setting these to their maximum values as well?

```
      .maxNestingDepth(Int.MaxValue)
      .maxNumberLength(Int.MaxValue)
```

Reply:
Good question. 
I am doubting a little disabling arbitrary limits.   

To be sure, I scanned `JSonProtocol` and I do not see any recursion which can lead to hitting `maxNestingDepth`. 
And I am also not seeing how numbers van have more than 1000 digits in the numbers written bij `JSonProtocol`.

So my advice would be not to do it.  But I am easily convinced otherwise.

Reply:
That's fine with me, thanks for clarifying!

Reply:
also cc @pjfanning

Reply:
`Int.MaxValue` limits the length a string can have to 2 gigabyte.

The reason this option was added to jackson was to give it a more predictable memory usage.  So that in a multi tenant system, you can fail fast when an attacker tries to send you 100 json documents to exhaust your memory and bring down the server.

But in terms of spark history server I think you can argue that spending worst case 2 gigabytes of buffers is acceptable as we do not have that kind of parallel load which can multiply this problem.

But again, I am easily convinced to make it configurable if you feel it is necessary.  But I prefer things to work out of the box.

Reply:
@LuciferYang the change looks ok to me. I would suggest that the Spark Team could consider introducing a new data format instead of storing this data as JSON. Parsing JSON with strings of up to 2Gb will take a lot of heap. Maybe at some stage, Jackson could support a streaming API like SAX where the char data can be broken up into chunks for large strings.

One question: what are you going to do when someone ends up with more than 2Gb of data stored in a big string?

I did notice an issue in JsonUtil. toJsonString wastes a huge amount of memory by not creating a Generator for a StringWriter. Creating on for a ByteArrayOutputSteam and then converting the byte array to a string means you need twice as much memory. You end up with the JSON as byte array and String in memory at same time. Probably is a reasonable amount slower too. Will I create an issue for this?